python/ray/data/tests/test_dynamic_block_split.py

```diff
@@ -0,0 +1,72 @@
+import numpy as np
+import pandas as pd
+import pytest
+
+import ray
+from ray.data.block import BlockMetadata
+from ray.data.datasource import Datasource
+from ray.data.datasource.datasource import ReadTask, Reader
+
+from ray.tests.conftest import *  # noqa
+
+
+def test_read_large_data(ray_start_cluster):
+    # Test 20G input with single task
+    num_batch = 20
+    ctx = ray.data.context.DatasetContext.get_current()
+    block_splitting_enabled = ctx.block_splitting_enabled
+    ctx.block_splitting_enabled = True
+
+    try:
+        cluster = ray_start_cluster
+        cluster.add_node(num_cpus=1)
+
+        ray.init(cluster.address)
+
+        # Data source generates multiple 1G random bytes data
+        class LargeBytesDatasource(Datasource):
+            def create_reader(self, **read_args):
+                return LargeBytesReader()
+
+        class LargeBytesReader(Reader):
+            def estimate_inmemory_data_size(self):
+                return None
+
+            def get_read_tasks(self, parallelism: int):
+                def _1g_batches_generator():
+                    for _ in range(num_batch):
+                        yield pd.DataFrame(
+                            {"one": [np.random.bytes(1024 * 1024 * 1024)]}
+                        )
+
+                return parallelism * [
+                    ReadTask(
+                        lambda: _1g_batches_generator(),
+                        BlockMetadata(
+                            num_rows=None,
+                            size_bytes=None,
+                            schema=None,
+                            input_files=None,
+                            exec_stats=None,
+                        ),
+                    )
+                ]
+
+        def foo(batch):
+            return pd.DataFrame({"one": [1]})
+
+        ds = ray.data.read_datasource(
+            LargeBytesDatasource(),
+            parallelism=1,
+        )
+
+        ds = ds.map_batches(foo, batch_size=None)
```

Comment:
Actually do you need to set a memory limit for the cluster? How does this make sure it will OOM without block splitting?

Reply:
After talking to Core folks, no easy way to set memory limit for the cluster in CI unit test now. As you can see here, we already process 20G data in one task. Planning to enable dynamic block splitting by default in 2.2. So I am not sure if it's our best interest to chase down having OOM without block splitting.